rotkehlchen/chain/ethereum/tokens.py

```diff
@@ -141,116 +156,59 @@ def query_tokens_for_addresses(
+                    account=account,
+                    chunks=get_chunks(all_tokens, n=ETHERSCAN_MAX_TOKEN_CHUNK_LENGTH),
+                    querying_etherscan=True,
+                )
+            self.db.save_tokens_for_address(write_cursor, account, list(token_amounts.keys()))
 
-    def _get_multitoken_account_balance(
+    def query_tokens_for_addresses(
             self,
-            tokens: List[EthereumToken],
-            account: ChecksumEthAddress,
-            call_order: Optional[Sequence[WeightedNode]],
-    ) -> Dict[str, FVal]:
-        """Queries balances of multiple tokens for an account
-
-        Return a dictionary with keys being tokens and value a dictionary of
-        account to balances
+            addresses: List[ChecksumEthAddress],
+    ) -> TokensReturn:
+        """Queries token balances for a list of addresses
+        Returns the token balances of each address and the usd prices of the tokens
 
         May raise:
         - RemoteError if an external service such as Etherscan is queried and
           there is a problem with its query.
         - BadFunctionCallOutput if a local node is used and the contract for the
           token has no code. That means the chain is not synced
         """
-        log.debug(
-            'Querying ethereum chain for multi token account balances',
-            eth_address=account,
-            tokens_num=len(tokens),
-        )
-        balances: Dict[str, FVal] = {}
-        result = ETH_SCAN.call(
-            ethereum=self.ethereum,
-            method_name='tokensBalance',
-            arguments=[account, [x.ethereum_address for x in tokens]],
-            call_order=call_order,
-        )
-        for tk_idx, token in enumerate(tokens):
-            token_amount = result[tk_idx]
-            if token_amount != 0:
-                normalized_amount = token_normalized_value(token_amount, token)
-                log.debug(
-                    f'Found {token.symbol}({token.ethereum_address}) token balance for '
-                    f'{account} and amount {normalized_amount}',
-                )
-                balances[token.identifier] = normalized_amount
-        return balances
+        token_usd_price: Dict[EthereumToken, Price] = {}
+        result = {}
+
+        for address in addresses:
+            with self.db.conn.read_ctx() as cursor:
+                saved_list = self.db.get_tokens_for_address(cursor, address=address)
+            if saved_list is None or len(saved_list) == 0:
+                continue  # Do not query if we know the address has no tokens
+
+            token_amounts = self._get_token_amounts(
+                account=address,
+                tokens=saved_list,
+                call_order=None,  # use defaults
+            )
+            result[address] = token_amounts
+
+            for token in token_amounts.keys() - token_usd_price:
+                token_usd_price[token] = Inquirer.find_usd_price(asset=token)
```

Comment:
This is an example of badly written code. There is too much happening and it's all implicit and requires the reader to know too many things. Good code is simple and readable.

1. You need to know that `keys()` returns a Dictionary view object that also happens to be set-like
2. You need to know that the `token_usd_price` gets coerced into being a set.
3. So in the end you have a set subtraction here.

The way to make it more readable is either assign `token_amounts.keys() - token_usd_price` into a variable named `tokens_without_price` and then iterate that variable or explain what you are doing above the loop with a comment.

`# Get the subset of tokens we don't yet know the price for`

------

But most importantly this is probably not needed.

`Inquirer.find_usd_price()` has a timed cache for the current price. Check its implementation. So the optimization you are doing is not required.

Also you should not query prices inside the loop. Instead gather all tokens at some point and query prices at the very end.

Reply:
Got it, done, thanks for the explanation